File: LeetCode/Medium/1208-get-equal-substrings-within-budget/1208-get-equal-substrings-within-budget.go
func equalSubstring(s string, t string, maxCost int) int {
    n := len(s)
    diff := make([]int, n)
    ret := 0
    l, curr := 0, 0

    for i := 0; i < n; i++ { diff[i] = abs(int(s[i]) - int(t[i])) }

    for r := 0; r < n; r++ {
        curr += diff[r]


        if curr > maxCost {
            curr -= diff[l]
            l++
        }

        ret = max(ret, r - l + 1)
    }

    return ret
}

func abs(a int) int {
    if a < 0 { return -a }
    return a
}

func max(a, b int) int {
    if a > b { return a }
    return b
}

File: LeetCode/Medium/1208-get-equal-substrings-within-budget/1208-get-equal-substrings-within-budget.py
class Solution:
    def equalSubstring(self, s: str, t: str, maxCost: int) -> int:
        diff = [abs(ord(s[i]) - ord(t[i])) for i in range(len(s))]
        ret = 0

        l, r, curr = 0, 0, 0

        while r < len(s):
            if curr + diff[r] <= maxCost:
                curr += diff[r]
                r += 1
                ret = max(ret, r - l)
            else:
                curr -= diff[l]
                l += 1
            
        return ret